[4.0] Improve updater layout

bembelimen opened the pull request:
## Summary of Changes
This PR improves the Joomla! updater layout and adds some structure:

### Initial layout
When we're in the initial state and did not check for the update yet, the following screen is visible:

#### Old
![grafik](https://user-images.githubusercontent.com/1229869/125176528-7a884780-e1d4-11eb-9f16-8978c6adb00d.png)

#### New
![grafik](https://user-images.githubusercontent.com/1229869/125176549-925fcb80-e1d4-11eb-82d3-daa353147c9f.png)

As you can see I used the emptystate function, which was implemented a few weeks ago (also in later stages). I also removed the information of the currently installed version, as it gives no benefit in the current state.

Last but not least the "Upload & Update" is moved down in a text link, as I assume upload updates are not that common.

### No update found
Now we press the "Check for updates" without finding an update.

#### Old
![grafik](https://user-images.githubusercontent.com/1229869/125176717-ffc02c00-e1d5-11eb-9f54-0438d6a2314f.png)

and

![grafik](https://user-images.githubusercontent.com/1229869/125176725-0babee00-e1d6-11eb-840a-6b8b330e767b.png)

#### New
![grafik](https://user-images.githubusercontent.com/1229869/125176737-28482600-e1d6-11eb-8486-e01ebbc232a7.png)

As you can see, there is no pre-update check, because....there is no update. But the button changes to offer the re-installation of the core files. Also all this unnecessary verbose information were removed.

### Update found
Next test is, if there is an update.

#### Old
![grafik](https://user-images.githubusercontent.com/1229869/125176894-9b05d100-e1d7-11eb-8e14-e986a2650c2d.png)

and

![grafik](https://user-images.githubusercontent.com/1229869/125176925-eb7d2e80-e1d7-11eb-8d56-0a18c08dbddd.png)

#### New
![grafik](https://user-images.githubusercontent.com/1229869/125176940-17001900-e1d8-11eb-9a0b-3ea5050a57fb.png)

and

![grafik](https://user-images.githubusercontent.com/1229869/125176944-20898100-e1d8-11eb-968f-c4180b509770.png)

and

![grafik](https://user-images.githubusercontent.com/1229869/125176948-2f703380-e1d8-11eb-9bee-777a09cb568c.png)

and

![grafik](https://user-images.githubusercontent.com/1229869/125176990-9988d880-e1d8-11eb-94ba-f272ec98c941.png)

Here the strange fieldsets are were removed and the information are split into different tab-pills. The update is moved to a 2nd step. Verbose information like current version and the download link are removed (no download needed here, see below in upload area). An additional checkbox is displayed for an extra confirmation regarding backups. Also the confirmation regarding extensions with update required is there.

The pills (left side) show an status icon, for extensions it's an animated hourclass till all extensions are checked.

### Update via upload
Last but not least the update via upload.

#### Old
![grafik](https://user-images.githubusercontent.com/1229869/125177127-9d692a80-e1d9-11eb-9a08-821f1491182e.png)

and

![grafik](https://user-images.githubusercontent.com/1229869/125177160-d0abb980-e1d9-11eb-8da4-83120dd830c0.png)

#### New
![grafik](https://user-images.githubusercontent.com/1229869/125177188-051f7580-e1da-11eb-9e82-a8eab2f8600f.png)

and

![grafik](https://user-images.githubusercontent.com/1229869/125177216-38620480-e1da-11eb-8bf1-f1fb3a96d9bb.png)

Here, if a update file is available as download, it's offered to the user to download. Headline of the Login is not "Login" anymore, because the user is already logged in. Also the backup confirmation is here again.

### Update
The update did not change.



### Testing Instructions
It's a bit hard to test. What I did was to install an [Joomla! RC2](https://github.com/joomla/joomla-cms/releases/tag/4.0.0-rc2) and replace the administrator/com_joomlaupdate and media/com_joomlaupdate folders from the download of this PR (see link at the bottom). Also the language file needs to be updated and layouts/joomla/content/emptystate.php has to be copied.

Then you can go through all the steps of the update.

For reinstall, you can use the normal patchtester.

## Actual result BEFORE applying this Pull Request
Update works but layoutit a bit confusing.


## Expected result AFTER applying this Pull Request
Update still works with new layout.

## Additional comment
Help pages has the wrong link.

## TODO
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877710241
- [ ] <s> https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877761125 </s> intended behaviour
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877761600
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877762033
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877762346
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877766531
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877762860
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877765407
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877781674
- [x] https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877714489

bembelimen committed:
First rework of update layouts

bembelimen committed:
Update precheck

bembelimen committed:
Fix order of empty pages

bembelimen committed:
Add reinstall and noupdate info screens

bembelimen committed:
Add update channel info

bembelimen committed:
Add update channel info link

bembelimen committed:
Add JS for (re-) install

bembelimen committed:
Submit update form

bembelimen committed:
fix broken query selector

bembelimen committed:
Remove useless if

bembelimen committed:
Fix notice if download data is not available

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into pre-updatecheck

bembelimen committed:
Add update confirm checkbox
Add update layout

bembelimen committed:
Add backup confirmation checkbox

bembelimen committed:
Add missing form

bembelimen committed:
Update pre-update-check view

bembelimen committed:
Add pre-update-check headline
Load installer language files
Add ignore extension checkbox

bembelimen committed:
Don't go to update view if critical PHP options are there
Change icon for extension to watch
Add some white background to login

bembelimen committed:
add preupdatecheck id to target JS

bembelimen committed:
Add warning messages

bembelimen committed:
Offer update download when available

bembelimen committed:
Add icon to precheck headlines
Update layout

bembelimen committed:
Move inline events to JS file
Hide precheck tables

bembelimen committed:
Move critical plugin criteria to model

bembelimen committed:
Add confirm box for upload
Cleaned up model call in view
Add version remark in update layout

bembelimen committed:
Add back button to update layout
Improve responsiveness

bembelimen committed:
Improve precheck infos
Rename upgrade => update

bembelimen committed:
Update JS
Add open/close toggler

bembelimen committed:
Make a string more personl
Fix ID

bembelimen committed:
Remove unused files

brianteeman commented:
Thanks for this. Looks a big improvement. I will take a proper look at this tomorrow and report back. Just one thing that strikes me right away (and to be fair its true without this PR but this PR makes it more obvious).

> When we're in the initial state and did not check for the update yet,

It says "no update available."

Maybe it should be saying "press this big button to see if there are any updates"

richard67 commented:
@bembelimen Please fix PHP CS: https://ci.joomla.org/joomla/joomla-cms/45704/1/6

richard67 commented:
@bembelimen The check for 8 MB or even 16 MB is outdated. A J4 update package currently has some 24 MB. A post_max_size or and upload_max_filesize lower than 24 MB will cause an update via the live update to fail. And maybe even 24 MB is too tight. In my opinion and according to my experience the 2 values which I have mentioned should both be at least 32 MB, with the post_max_size being equal to or larger than the upload_max_filesize, and the memory limit should be at least the maximum of these 2 values, if not more.

brianteeman commented:
Warning: Creating default object from empty value in C:\htdocs\joomla-cms\administrator\components\com_joomlaupdate\src\Model\UpdateModel.php on line 1336
![image](https://user-images.githubusercontent.com/1296369/125187781-31aeae00-e229-11eb-8b4d-51bc297e8870.png)

brianteeman commented:
Couldnt get to the re-install core package page

richard67 commented:
@brianteeman Both code line you've mentioned haven't been changed by this PR, so the mistake seems to happen before that in the call stack. Can you switch on "Debug System" in Global Configuration so you can see the call stack?

brianteeman commented:
icon alignment

![image](https://user-images.githubusercontent.com/1296369/125187982-25772080-e22a-11eb-86d2-51960ea37efc.png)

brianteeman commented:
Click to uncheck the "ignore warning" a second time (to uncheck it)

`Uncaught TypeError: e.addAttribute is not a function
    at default.min.js?21f709b9142f3119910d9af57d0d8814:1
    at Array.forEach (<anonymous>)
    at HTMLInputElement.l (default.min.js?21f709b9142f3119910d9af57d0d8814:1)
(anonymous) @ default.min.js?21f709b9142f3119910d9af57d0d8814:1
l @ default.min.js?21f709b9142f3119910d9af57d0d8814:1`

![image](https://user-images.githubusercontent.com/1296369/125188132-96b6d380-e22a-11eb-858c-6243f8d11e78.png)

brianteeman commented:
There is a popover :( that is not visually indicated and is not accessible from the keyboard. Oh and the text needs to be updated

![image](https://user-images.githubusercontent.com/1296369/125188234-fe6d1e80-e22a-11eb-9f27-4ccb18bbc532.png)

brianteeman commented:
The lack of styling on the tables means that the "more details" toggle is _lost_ on the page and its not clear what it refers to

![image](https://user-images.githubusercontent.com/1296369/125188282-38d6bb80-e22b-11eb-8494-6510eec404cd.png)

brianteeman commented:
After completing an update you are taken to the url `administrator/index.php?option=com_joomlaupdate&view=joomlaupdate&layout=complete` but the on screen text doesnt indicate that you have updated

![image](https://user-images.githubusercontent.com/1296369/125188430-aa166e80-e22b-11eb-9c21-eaf0ed0d6804.png)

brianteeman commented:
All tables must have a caption that describes the table.  
table headers must have a scope=col
The extension name should be a table header with a scope=row

![image](https://user-images.githubusercontent.com/1296369/125189050-43468480-e22e-11eb-8eb7-b58b189e78fe.png)

brianteeman commented:
When an update has been found - clicking on the "check for updates" button in the toolbar produces a js error

`core.js?c47dc50c9a8b9c85d135bbac3f47a6e9b4100832:389 Uncaught TypeError: Cannot read property 'task' of null
    at Object.Joomla.submitform (core.js?c47dc50c9a8b9c85d135bbac3f47a6e9b4100832:389)
    at Object.Joomla.submitbutton (core.js?c47dc50c9a8b9c85d135bbac3f47a6e9b4100832:446)
    at HTMLElement.executeTask (joomla-toolbar-button.js?176b6426ebbc64fc3b01addcab4f31ae384376d0:120)
Joomla.submitform @ core.js?c47dc50c9a8b9c85d135bbac3f47a6e9b4100832:389
Joomla.submitbutton @ core.js?c47dc50c9a8b9c85d135bbac3f47a6e9b4100832:446
executeTask @ joomla-toolbar-button.js?176b6426ebbc64fc3b01addcab4f31ae384376d0:120`

brianteeman commented:
> Creating default object from empty value 

This is a different bug from #34115 - I will raise a separate issue for that

bembelimen commented:
Thanks for the feedback so far. I'll look into it over the next days.
Two things to mention:
- the language strings are somehow "strange" english (not sure if they just look like that for me, as I'm not native or someone non-native created them), so probably worth reviewing them (I also like the idea of your suggestion replacing "no update available").
- With this PR I tried to not touch any logic in terms of model thing but it seems, I should do that (also stuff like moving the DB query in the view into the model).

Thanks again for the feedback, will try to fix everything.

brianteeman commented:
the language strings are indeed strange - its not you

rinka88 commented:
it would be great to see it in an upcoming release candidate

brianteeman commented:
please add the deleted files to the list in script.php

brianteeman commented on administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php:
```diff
@@ -81,42 +117,81 @@ class HtmlView extends BaseHtmlView
-			ToolbarHelper::preferences('com_joomlaupdate');
+			$this->setLayout('selfupdate');
+		}
+		// User requests the manual update and is an admin
```
```suggestion
		// User requests the manual update and is a super user
```

brianteeman commented on administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php:
```diff
@@ -81,42 +117,81 @@ class HtmlView extends BaseHtmlView
 		}
+		elseif (!$hasDownload || !$hasUpdate)
+		{
+			// Could be that we have a download file but no update, so we offer an re-install
```
```suggestion
			// Could be that we have a download file but no update, so we offer a re-install
```

brianteeman commented on administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php:
```diff
@@ -81,42 +117,81 @@ class HtmlView extends BaseHtmlView
+			// Could be that we have a download file but no update, so we offer an re-install
+			if ($hasDownload)
+			{
+				// We can reinstall if we have an URL but no update
```
```suggestion
				// We can reinstall if we have a URL but no update
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/noupdate.php:
```diff
@@ -0,0 +1,49 @@
+	$displayData['content'] .= '<br>' . HTMLHelper::_('link',
+		$this->updateInfo['object']->get('infourl')->_data,
+		Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL'),
+		[
```
```suggestion
			[
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,344 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_joomlaupdate
+ *
+ * @copyright   (C) 2005 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,344 @@
+	->useScript('bootstrap.popover')
+	->useScript('bootstrap.tab');
+
+// JText::script doesn't have a sprintf equivalent so work around this
```
```suggestion
// Text::script doesn't have a sprintf equivalent so work around this
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/reinstall.php:
```diff
@@ -0,0 +1,57 @@
+	$displayData['content'] .= '<br>' . HTMLHelper::_('link',
+		$this->updateInfo['object']->get('infourl')->_data,
+		Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL'),
+		[
```
```suggestion
			[
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/selfupdate.php:
```diff
@@ -0,0 +1,23 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_joomlaupdate
+ *
+ * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/upload.php:
```diff
@@ -0,0 +1,100 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_joomlaupdate
+ *
+ * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

wilsonge commented:
Just for transparency here - as we're refactoring strings as well here - I've discussed with @bembelimen and we agreed that either this goes in time for RC4 on Tuesday or it will get delayed for 4.1 stable. Obviously this is a massive improvement so it will get in one way or another.

bembelimen commented:
> 
> 
> Couldnt get to the re-install core package page

Could it be, that this notice is responsible that the reinstall does not work? I neither get the notice nor can't I re-install it. So if I click the button, I come to the re-install screen and get all files resetted.

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/selfupdate.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/upload.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented:
>> Couldnt get to the re-install core package page

> Could it be, that this notice is responsible that the reinstall does not work? 

No the notice was something else and unrelated. I think its something to do with the update servers or version?? See animation below.

![update](https://user-images.githubusercontent.com/1296369/125212165-81ca5680-e2a3-11eb-8972-6961615b749c.gif)

bembelimen committed:
Re-add complete message

bembelimen committed:
Move self check to model
Move uploadcheck to upload view
Rename first info headline when no update is loaded

bembelimen committed:
Fix JS error
Uncheck checkbox if user cancel confirm box

bembelimen committed:
Add form to allow update check in pre-check view

bembelimen committed:
Add files to delete list

bembelimen committed:
Merge branch '4.0/pre-updatecheck' of github.com:bembelimen/joomla-cms into 4.0/pre-updatecheck

# Conflicts:
#	administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php

bembelimen commented:
> 
> 
> > > Couldnt get to the re-install core package page
> 
> > Could it be, that this notice is responsible that the reinstall does not work?
> 
> No the notice was something else and unrelated. I think its something to do with the update servers or version?? See animation below.
> 
> ![update](https://user-images.githubusercontent.com/1296369/125212165-81ca5680-e2a3-11eb-8972-6961615b749c.gif)

Thanks for the screencast, now I can reproduce it and the explaination is very "easy" but also hard to solve: when you're at the default update channel, there is no version you can use (read: no link to rewrite the files). Same happens for the old version of the updater, so not really a bug here.

bembelimen committed:
Update min filesize to 32MB

bembelimen committed:
Remove comment

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/noupdate.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/reinstall.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
lint fix

bembelimen committed:
Merge branch '4.0/pre-updatecheck' of github.com:bembelimen/joomla-cms into 4.0/pre-updatecheck

bembelimen committed:
lint fix

bembelimen committed:
Move table description to caption
Add scope="col" to header

bembelimen commented:
> 
> 
> icon alignment
> 
> ![image](https://user-images.githubusercontent.com/1296369/125187982-25772080-e22a-11eb-86d2-51960ea37efc.png)

TBH I wouldn't change it here, as the icons have different widths... so to make them all vertical aligned I have to fiddle around with px and it only works till someone changes the icons...so they're currently just right aligned

bembelimen commented:
> 
> 
> There is a popover :( that is not visually indicated and is not accessible from the keyboard. Oh and the text needs to be updated
> 
> ![image](https://user-images.githubusercontent.com/1296369/125188234-fe6d1e80-e22a-11eb-9f27-4ccb18bbc532.png)

Yeah, that's really ugly + via JS injected. I have no real idea how to improve it beside removing it at all... Could you attach the plugin you used to test, so I can test myself that way?

And would you be able to look through the language files and fix the biggest issues? Probably we get it in 4.0, if we're fast...

bembelimen commented:
> 
> 
> The lack of styling on the tables means that the "more details" toggle is _lost_ on the page and its not clear what it refers to
> 
> ![image](https://user-images.githubusercontent.com/1296369/125188282-38d6bb80-e22b-11eb-8494-6510eec404cd.png)

No with the description moved to the caption, is the show more close enough?

hans2103 commented:
> > There is a popover :( that is not visually indicated and is not accessible from the keyboard. Oh and the text needs to be updated
> > ![image](https://user-images.githubusercontent.com/1296369/125188234-fe6d1e80-e22a-11eb-9f27-4ccb18bbc532.png)
> 
> Yeah, that's really ugly + via JS injected. I have no real idea how to improve it beside removing it at all... Could you attach the plugin you used to test, so I can test myself that way?
> 
> And would you be able to look through the language files and fix the biggest issues? Probably we get it in 4.0, if we're fast...

The mentioned plugin is inside PR https://github.com/joomla/joomla-cms/pull/34743

brianteeman commented:
> TBH I wouldn't change it here, as the icons have different widths... so to make them all vertical aligned I have to fiddle around with px and it only works till someone changes the icons...so they're currently just right aligned

There is a specific css class to use on the icons to make them fixed width

`.fa-fw, .icon-fw `

brianteeman commented on administrator/components/com_admin/script.php:
```diff
@@ -1395,6 +1395,14 @@ public function deleteUnexistingFiles($dryRun = false, $suppressOutput = false)
 			'/administrator/components/com_joomlaupdate/helpers/select.php',
 			'/administrator/components/com_joomlaupdate/joomlaupdate.php',
 			'/administrator/components/com_joomlaupdate/models/default.php',
+			'/administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default.php',
```
This list is grouped by version as well as sorted alphabetically so these files need to be moved to the section beginning `// 4.0 from RC 3 to RC 4` around line 5870

richard67 replied:
No, they need to go to section "// From 3.10 to 4.0" because sections are not named by when the code was added to script.php but by the versions which are compared with the tool.

If after the removal of the files the parent folder(s) are empty, the folders need to be removed, too. Folder list has same sections but is ordered in reverse alphabetical order.

richard67 replied:
@wilsonge If I get pinged after this PR is merged but before RC 4 is built, I can run the tool to generate the lists and make a quick PR for updating script.php.

brianteeman commented on administrator/components/com_joomlaupdate/src/Controller/DisplayController.php:
```diff
@@ -45,6 +45,12 @@ public function display($cachable = false, $urlparams = false)
 		// Get and render the view.
 		if ($view = $this->getView($vName, $vFormat))
 		{
+			// Only super user can access file upload
```
I thought only a super user could access the entire component? If so then is this additional check needed?

bembelimen replied:
That's true [for the backend](https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_joomlaupdate/src/Dispatcher/Dispatcher.php#L32) (I'm not sure why the check is the way it is). So e.g. in a CLI context one could access without core.admin access. I just took over the permission check.

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+$compatibilityTypes = array(
+	'COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_RUNNING_PRE_UPDATE_CHECKS' => array(
+		'class' => 'info',
+		'icon' => 'hourglass fa-spin',
```
```suggestion
		'icon'  => 'hourglass fa-spin',
```

very fussy but we do align the => elsewhere

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+	),
+	'COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_REQUIRING_UPDATES_TO_BE_COMPATIBLE' => array(
+		'class' => 'danger',
+		'icon' => 'times',
```
```suggestion
		'icon'  => 'times',
```
very fussy but we do align the => elsewhere

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+	),
+	'COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PRE_UPDATE_CHECKS_FAILED' => array(
+		'class' => 'warning',
+		'icon' => 'exclamation-triangle',
```
```suggestion
		'icon'  => 'exclamation-triangle',
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+	),
+	'COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_UPDATE_SERVER_OFFERS_NO_COMPATIBLE_VERSION' => array(
+		'class' => 'warning',
+		'icon' => 'exclamation-triangle',
```
```suggestion
		'icon'  => 'exclamation-triangle',
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+	),
+	'COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_PROBABLY_COMPATIBLE' => array(
+		'class' => 'success',
+		'icon' => 'check',
```
```suggestion
		'icon'  => 'check',
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+
+<div id="joomlaupdate-wrapper" class="main-card p-3 mt-3" data-joomla-target-version="<?php echo $latestJoomlaVersion; ?>" data-joomla-current-version="<?php echo $currentJoomlaVersion; ?>">
+
+	<h2 class="mt-3 mb-3">
```
```suggestion
	<h2 class="my-3">
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/upload/default.php:
```diff
@@ -0,0 +1,100 @@
+	<div class="form-check mb-3">
+		<input class="form-check-input me-2" type="checkbox" value="" id="joomlaupdate-confirm-backup">
+		<label class="form-check-label" for="joomlaupdate-confirm-backup">
+		<?php echo Text::_('COM_JOOMLAUPDATE_UPDATE_CONFIRM_BACKUP'); ?>
```
```suggestion
			<?php echo Text::_('COM_JOOMLAUPDATE_UPDATE_CONFIRM_BACKUP'); ?>
```

brianteeman commented on layouts/joomla/content/emptystate.php:
```diff
@@ -45,7 +45,7 @@
 				<?php if ($createURL && Factory::getApplication()->input->get('tmpl') !== 'component') : ?>
 					<a href="<?php echo Route::_($createURL); ?>"
-					   class="btn btn-primary btn-lg px-4 me-sm-3"><?php echo $btnadd; ?></a>
+					   class="btn btn-primary btn-lg px-4 me-sm-3 emptystate-btnadd"><?php echo $btnadd; ?></a>
```
What is the purpose of adding this class - it doesnt exist in the css

bembelimen replied:
It's needed here: https://github.com/joomla/joomla-cms/pull/34754/files#diff-550b2e9757e1fef80e3714191e4f70f3dd9bd7441ff3e7679698de031551e346R49

To submit the form I have to recognize the click event on this button....and there was no way to distinguish it from the others.

brianteeman replied:
would an id not be better?

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+						<?php break; ?>
+					<?php endif; ?>
+				<?php endforeach; ?>
+				<span class="fa fa-<?php echo $labelClass == 'danger' ? 'times' : 'check'; ?> p-1 bg-white ms-2 text-<?php echo $labelClass; ?>"></span>
```
```suggestion
				<span class="fa fa-<?php echo $labelClass == 'danger' ? 'times' : 'check'; ?> p-1 bg-white ms-2 text-<?php echo $labelClass; ?>" aria-hidden="true"></span>
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+						<?php break; ?>
+					<?php endif; ?>
+				<?php endforeach; ?>
+				<span class="fa fa-<?php echo $labelClass == 'warning' ? 'exclamation-triangle' : 'check'; ?> p-1 bg-white ms-2 text-<?php echo $labelClass; ?>"></span>
```
```suggestion
				<span class="fa fa-<?php echo $labelClass == 'warning' ? 'exclamation-triangle' : 'check'; ?> p-1 bg-white ms-2 text-<?php echo $labelClass; ?>" aria-hidden="true"></span>
```

bembelimen committed:
Make icons fixed width

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+			<button class="nav-link d-flex justify-content-between align-items-center" id="joomlaupdate-precheck-extensions-tab" data-bs-toggle="pill" data-bs-target="#joomlaupdate-precheck-extensions-content" type="button" role="tab" aria-controls="joomlaupdate-precheck-extensions-content" aria-selected="false">
+				<?php echo Text::_('COM_JOOMLAUPDATE_PREUPDATE_EXTENSIONS'); ?>
+				<?php $labelClass = 'success'; ?>
+				<span class="fa fa-hourglass fa-spin p-1 bg-white ms-2 text-info"></span>
```
```suggestion
				<span class="fa fa-hourglass fa-spin p-1 bg-white ms-2 text-info" aria-hidden="true"></span>
```

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+								<button type="button" class="btn btn-link compatibilitytoggle ms-auto" data-state="closed">
+									<?php echo Text::sprintf(
+										'COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_MORE_COMPATIBILITY_INFORMATION',
+										'<span class="icon-chevron-right"></span>'
```
```suggestion
										'<span class="icon-chevron-right" aria-hidden="true"></span>'
```

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+							<table class="table table-striped">
+								<thead class="row-fluid">
+									<tr>
+										<th class="exname">
+											<?php echo Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_NAME'); ?>
+										</th>
+										<th class="extype">
+											<?php echo Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_TYPE'); ?>
+										</th>
+										<th class="instver hidden">
+											<?php echo Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_INSTALLED_VERSION'); ?>
+										</th>
+										<th class="upcomp hidden">
+											<?php echo Text::sprintf('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_COMPATIBLE_WITH_JOOMLA_VERSION', isset($this->updateInfo['current']) ? $this->escape($this->updateInfo['current']) : JVERSION); ?>
+										</th>
+										<th class="currcomp hidden">
+											<?php echo Text::sprintf('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_COMPATIBLE_WITH_JOOMLA_VERSION', $this->escape($this->updateInfo['latest'])); ?>
+										</th>
```
```suggestion
										<th scope="col">
											<?php echo Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_NAME'); ?>
										</th>
										<th scope="col">
											<?php echo Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_TYPE'); ?>
										</th>
										<th class="hidden" scope="col">
											<?php echo Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_INSTALLED_VERSION'); ?>
										</th>
										<th class="hidden" scope="col">
											<?php echo Text::sprintf('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_COMPATIBLE_WITH_JOOMLA_VERSION', isset($this->updateInfo['current']) ? $this->escape($this->updateInfo['current']) : JVERSION); ?>
										</th>
										<th class="hidden" scope="col">
											<?php echo Text::sprintf('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_COMPATIBLE_WITH_JOOMLA_VERSION', $this->escape($this->updateInfo['latest'])); ?>
										</th>
```

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/upload/default.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+								<?php if ($data['group'] == 0) : ?>
+									<?php foreach ($this->nonCoreExtensions as $extension) : ?>
+										<tr>
+											<td class="exname">
+												<?php echo $extension->name; ?>
+											</td>
+											<td class="extype">
+												<?php echo Text::_('COM_INSTALLER_TYPE_' . strtoupper($extension->type)); ?>
+											</td>
+											<td class="instver hidden">
+												<?php echo $extension->version; ?>
+											</td>
+											<td id="available-version-<?php echo $extension->extension_id; ?>" class="currcomp hidden" />
+											<td
+												class="extension-check upcomp hidden"
+												data-extension-id="<?php echo $extension->extension_id; ?>"
+												data-extension-current-version="<?php echo $extension->version; ?>"
+											>
+												<img src="<?php echo Uri::root(true); ?>/media/system/images/ajax-loader.gif">
+											</td>
```
```suggestion
											<th scope="row">
												<?php echo $extension->name; ?>
											</td>
											<td>
												<?php echo Text::_('COM_INSTALLER_TYPE_' . strtoupper($extension->type)); ?>
											</td>
											<td class="hidden">
												<?php echo $extension->version; ?>
											</td>
											<td id="available-version-<?php echo $extension->extension_id; ?>" class="hidden" />
											<td
												class="hidden"
												data-extension-id="<?php echo $extension->extension_id; ?>"
												data-extension-current-version="<?php echo $extension->version; ?>"
											>
												<img src="<?php echo Uri::root(true); ?>/media/system/images/ajax-loader.gif">
											</td>
```

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php:
```diff
@@ -0,0 +1,349 @@
+
+		<div id="preupdatecheckbox">
+			<div class="form-check d-flex justify-content-center mb-3">
+				<input type="checkbox" class="me-3" id="noncoreplugins" name="noncoreplugins" value="1" required aria-required="true" />
```
```suggestion
				<input type="checkbox" class="me-3" id="noncoreplugins" name="noncoreplugins" value="1" required />
```

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented:
The icon animation checks with an hourglass are different to similar icon animations in the system dashboard. It would be good if they could be updated here to match

![spin2](https://user-images.githubusercontent.com/1296369/125254436-5f135e80-e2f2-11eb-8d41-9f194d7dd909.gif)

![spin](https://user-images.githubusercontent.com/1296369/125254574-810ce100-e2f2-11eb-982a-8549f3c14634.gif)

bembelimen committed:
Merge branch '4.0/pre-updatecheck' of github.com:bembelimen/joomla-cms into 4.0/pre-updatecheck

# Conflicts:
#	administrator/components/com_joomlaupdate/tmpl/joomlaupdate/preupdatecheck.php

bembelimen commented:
> 
> 
> > TBH I wouldn't change it here, as the icons have different widths... so to make them all vertical aligned I have to fiddle around with px and it only works till someone changes the icons...so they're currently just right aligned
> 
> There is a specific css class to use on the icons to make them fixed width
> 
> `.fa-fw, .icon-fw `

Thx

bembelimen committed:
Change hourglass icon to spinner

bembelimen commented:
> 
> 
> The icon animation checks with an hourglass are different to similar icon animations in the system dashboard. It would be good if they could be updated here to match
> 
> ![spin2](https://user-images.githubusercontent.com/1296369/125254436-5f135e80-e2f2-11eb-8d41-9f194d7dd909.gif)

Done

brianteeman commented:
_(offline for the next few hours for media interviews - will check again (inc lang) this afternoon)_

bembelimen committed:
Fix spinner icon

richard67 commented:
> This list is grouped by version as well as sorted alphabetically so these files need to be moved to the section beginning `// 4.0 from RC 3 to RC 4` around line 5870

@brianteeman That's not right. The PR is good as it is regarding script.php. The need to go to section "// From 3.10 to 4.0" because sections are not named by when the code was added to script.php but by the versions which are compared with the tool, and that's what this PR does.

brianteeman commented:
Which is why it needs to be moved as the files were removed after rc3 and not after 3. 10

Otherwise it means all the files are in the wrong place

bembelimen committed:
Improve popover + warning of critical plugins

richard67 commented:
> Which is why it needs to be moved as the files were removed after rc3 and not after 3. 10

@brianteeman No, you still don't get it. The sections are based on the versions of full packages which are compared by the tool, not on then something was added to script.php. If something is removed when updating from 3.10 to 4, like it is here, it belongs to the section "// From 3.10 to 4.0". Syy all my recent PR's for script.php where I have already moved other stuff.

> Otherwise it means all the files are in the wrong place

No, nothing is at the wrong place right now in script.php.

brianteeman commented:
But it's also removed when updating from RC3

richard67 commented:
> But it's also removed when updating from RC3

@brianteeman Correct, and so it would be added to the script 2 times. The improved version of the tool which I used in recent times doesn't add something to a later section if it is already present in the "// From 3.10 to 4.0" so we don't have duplicates.

brianteeman commented:
where is this change in script.php documented

richard67 commented:
> where is this change in script.php documented

@brianteeman Nowhere right now in documents, but in all those recent PR's I had made for script.php it is described.

chmst commented:
![grafik](https://user-images.githubusercontent.com/1035262/125283769-03a89700-e319-11eb-9600-a7a9e5910bb0.png)

This is not a question.

chmst commented:
![grafik](https://user-images.githubusercontent.com/1035262/125285080-8bdb6c00-e31a-11eb-9475-739436ef1e90.png)
The upload file uses the defaul language of the site, not the user's language

bembelimen committed:
Convert question to statement

bembelimen committed:
Enable update button when no extensions are available

bembelimen commented:
> 
> 
> ![grafik](https://user-images.githubusercontent.com/1035262/125285080-8bdb6c00-e31a-11eb-9475-739436ef1e90.png)
> The upload file uses the defaul language of the site, not the user's language

You mean: "Keine Datei ausgewählt"? That's from the browser, as it's a normal upload field.

chmst commented:
Contrast errors on "badge bg-warning". But this is not related to this PR. 

Contrast error also here: Colour yellow on white: 

![grafik](https://user-images.githubusercontent.com/1035262/125304355-d2869180-e32d-11eb-8140-93c91a3470b3.png)

The buttons "More Details" are at the right side of the screen could be close to the information? 
Or make the whole information a button? 

Suggestion 

![grafik](https://user-images.githubusercontent.com/1035262/125307851-b3d5ca00-e330-11eb-851d-37633c046d4c.png)

brianteeman commented:
We use a very similar visual display for "permissions" which uses webcomponent **joomla-tab** shouldnt that be used here as well

brianteeman commented:
> @brianteeman Nowhere right now in documents, but in all those recent PR's I had made for script.php it is described.

Thats a really hard fail.

richard67 commented:
> 
> 
> > @brianteeman Nowhere right now in documents, but in all those recent PR's I had made for script.php it is described.
> 
> Thats a really hard fail.

@brianteeman Do we have any documentation already telling how script.php needs to be updated? As far as I know not. So tell me where have I failed?

Quy commented on administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php:
```diff
@@ -151,96 +204,48 @@ public function display($tpl = null)
-			// Always force to check for an update!
-			$cache_timeout = 0;
-
-			$updater = Updater::getInstance();
```
Please delete:

```
use Joomla\CMS\Updater\Updater;
use Joomla\Database\ParameterType;
```

brianteeman commented:
the behaviour has changed. before this undocumented change (even a comment in the code) I just needed to add a removed file to the list at the end. ie I only needed to know the version of joomla that the file is being removed from. Now I appear to also need to know which version of joomla that the file was created in

Quy commented on administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php:
```diff
@@ -151,96 +204,48 @@ public function display($tpl = null)
 		{
-			// Get the component extension ID
-			$joomlaUpdateComponentId = $db->loadResult();
+			$arrow  = Factory::getLanguage()->isRtl() ? 'arrow-right' : 'arrow-left';
```
```suggestion
			$arrow = Factory::getLanguage()->isRtl() ? 'arrow-right' : 'arrow-left';
```

Quy commented on administrator/components/com_joomlaupdate/src/View/Upload/HtmlView.php:
```diff
@@ -34,17 +61,40 @@ class HtmlView extends BaseHtmlView
+		// Set the toolbar information.
+		ToolbarHelper::title(Text::_('COM_JOOMLAUPDATE_OVERVIEW'), 'sync install');
+
+		$arrow  = Factory::getLanguage()->isRtl() ? 'arrow-right' : 'arrow-left';
```
```suggestion
		$arrow = Factory::getLanguage()->isRtl() ? 'arrow-right' : 'arrow-left';
```

Quy commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/noupdate.php:
```diff
@@ -0,0 +1,49 @@
+	$displayData['content'] .= '<br>' . HTMLHelper::_('link',
+		$this->updateInfo['object']->get('infourl')->_data,
+		Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL'),
+			[
+				'target' => '_blank',
+				'rel'    => 'noopener noreferrer',
+				'title'  => isset($this->updateInfo['object']->get('infourl')->title) ? Text::sprintf('JBROWSERTARGET_NEW_TITLE', $this->updateInfo['object']->get('infourl')->title) : ''
+			]
```
```suggestion
		[
			'target' => '_blank',
			'rel'    => 'noopener noreferrer',
			'title'  => isset($this->updateInfo['object']->get('infourl')->title) ? Text::sprintf('JBROWSERTARGET_NEW_TITLE', $this->updateInfo['object']->get('infourl')->title) : ''
		]
```

Quy commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/reinstall.php:
```diff
@@ -0,0 +1,57 @@
+	$displayData['content'] .= '<br>' . HTMLHelper::_('link',
+		$this->updateInfo['object']->get('infourl')->_data,
+		Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL'),
+			[
+				'target' => '_blank',
+				'rel'    => 'noopener noreferrer',
+				'title'  => isset($this->updateInfo['object']->get('infourl')->title) ? Text::sprintf('JBROWSERTARGET_NEW_TITLE', $this->updateInfo['object']->get('infourl')->title) : ''
+			]
```
```suggestion
		[
			'target' => '_blank',
			'rel'    => 'noopener noreferrer',
			'title'  => isset($this->updateInfo['object']->get('infourl')->title) ? Text::sprintf('JBROWSERTARGET_NEW_TITLE', $this->updateInfo['object']->get('infourl')->title) : ''
		]
```

Quy commented on administrator/components/com_joomlaupdate/tmpl/joomlaupdate/update.php:
```diff
@@ -0,0 +1,66 @@
+	$displayData['content'] .= '<br>' . HTMLHelper::_('link',
+		$this->updateInfo['object']->get('infourl')->_data,
+		Text::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL'),
+		[
+				'target' => '_blank',
+				'rel'    => 'noopener noreferrer',
+				'title'  => isset($this->updateInfo['object']->get('infourl')->title) ? Text::sprintf('JBROWSERTARGET_NEW_TITLE', $this->updateInfo['object']->get('infourl')->title) : ''
+			]
```
```suggestion
		[
			'target' => '_blank',
			'rel'    => 'noopener noreferrer',
			'title'  => isset($this->updateInfo['object']->get('infourl')->title) ? Text::sprintf('JBROWSERTARGET_NEW_TITLE', $this->updateInfo['object']->get('infourl')->title) : ''
		]
```

richard67 commented:
> the behaviour has changed. before this undocumented change (even a comment in the code) I just needed to add a removed file to the list at the end. ie I only needed to know the version of joomla that the file is being removed from. Now I appear to also need to know which version of joomla that the file was created in

@brianteeman The point is that for J4 it has once been agreed that maintaining the list of deleted files and folders can not be left to the authors of PR's since it has become more complicated than in J3 to determine which files or folder to be added, e.g. due to the compilation of resources in J4 (build/media_source). That's why it has once been decided that the release lead runs a tool which currently is under development here https://github.com/joomla/joomla-cms/pull/25559 , and because the release lead has lots of other stuff to do, too, when preparing a release, I took over that job from him and generate the lists before each RC release in recent times. So in general it is wrong that you require it to be done in PR's.

@wilsonge Please confirm.

brianteeman commented:
and my point again is that this change is not documented at all and that as a result of this changed policy the updating of the list of files to remove should **not** be included in the PR at all.

Quy commented on administrator/components/com_installer/src/Model/WarningsModel.php:
```diff
@@ -95,7 +95,11 @@ public function getItems()
-		$messages = array();
+		$messages = [];
+
+		 // 32MB
```
```suggestion
		// 32MB
```

Quy commented on administrator/components/com_installer/src/Model/WarningsModel.php:
```diff
@@ -127,24 +128,21 @@ public function getItems()
-		if ($memory_limit < (8 * 1024 * 1024) && $memory_limit != -1)
+		if ($memory_limit < $minMemory && $memory_limit != -1)
 		{
 			// 8MB
```
Is 8MB or 32MB per line 101?

bembelimen replied:
Set to 16/24MB as lower limits, but the language strings used here are also a f*** up.

Quy commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -15,26 +16,45 @@ COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_NEXT="Joomla Next"
+COM_JOOMLAUPDATE_EMPTYSTATE_APPEND="Update your site by uploading the update package manually."
+COM_JOOMLAUPDATE_EMPTYSTATE_BUTTON_ADD="Check for updates"
+COM_JOOMLAUPDATE_EMPTYSTATE_CONTENT="Click the button below to check for updates."
+COM_JOOMLAUPDATE_EMPTYSTATE_TITLE="Please check if an update is available."
```
```suggestion
COM_JOOMLAUPDATE_EMPTYSTATE_TITLE="Check if an update is available."
```

brianteeman commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -15,26 +16,45 @@ COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_NEXT="Joomla Next"
 COM_JOOMLAUPDATE_CONFIRM="Confirm"
+COM_JOOMLAUPDATE_EMPTYSTATE_APPEND="Update your site by uploading the update package manually."
+COM_JOOMLAUPDATE_EMPTYSTATE_BUTTON_ADD="Check for updates"
+COM_JOOMLAUPDATE_EMPTYSTATE_CONTENT="Click the button below to check for updates."
```
```suggestion
COM_JOOMLAUPDATE_EMPTYSTATE_CONTENT="Select the button below to check for updates."
```

you can't click with a finger or screen reader ;)

brianteeman commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -15,26 +16,45 @@ COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_NEXT="Joomla Next"
 COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_TESTING="Testing"
 COM_JOOMLAUPDATE_CONFIGURATION="Joomla Update: Options"
 COM_JOOMLAUPDATE_CONFIRM="Confirm"
+COM_JOOMLAUPDATE_EMPTYSTATE_APPEND="Update your site by uploading the update package manually."
```
```suggestion
COM_JOOMLAUPDATE_EMPTYSTATE_APPEND="Update your site by manually uploading the update package."
```

brianteeman commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -15,26 +16,45 @@ COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_NEXT="Joomla Next"
 COM_JOOMLAUPDATE_MINIMUM_STABILITY_LABEL="Minimum Stability"
 COM_JOOMLAUPDATE_MINIMUM_STABILITY_RC="Release Candidate"
 COM_JOOMLAUPDATE_MINIMUM_STABILITY_STABLE="Stable"
+COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_APPEND="Update your site by uploading the update package manually."
```
```suggestion
COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_APPEND="Update your site by manually uploading the update package."
```

brianteeman commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -15,26 +16,45 @@ COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_NEXT="Joomla Next"
+COM_JOOMLAUPDATE_REINSTALL_EMPTYSTATE_CONTENT="You can reinstall all core files to repair broken or missing files."
+COM_JOOMLAUPDATE_REINSTALL_EMPTYSTATE_TITLE="Currently no update available."
+COM_JOOMLAUPDATE_SELF_EMPTYSTATE_BUTTON_ADD="Update component now"
+COM_JOOMLAUPDATE_SELF_EMPTYSTATE_CONTENT="There is a new version of the Joomla Update Component available. You must update the Joomla Update Component first before you can update Joomla!"
```
```suggestion
COM_JOOMLAUPDATE_SELF_EMPTYSTATE_CONTENT="A new version of the Joomla Update Component is available. You must update this first before you can update Joomla!"
```

brianteeman commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -87,24 +107,25 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL="Additional Information"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_BEING_CHECKED="The system is currently checking these plugins to see if they could cause problems during the upgrade.<br><br>Please be patient while the checks are completed."
-COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="Do you wish to ignore the warnings about potentially incompatible plugins and to proceed with the upgrade?"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_BEING_CHECKED="The system is currently checking these plugins to see if they could cause problems during the update.<br><br>Please be patient while the checks are completed."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="I want to ignore the warnings about potentially incompatible extensions and to proceed with the update!"
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_PLUGIN_CONFIRMATION="I accept the warnings about potentially incompatible extensions and wish to proceed with the update."
```

brianteeman commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -87,24 +107,25 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL="Additional Information"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLAGAIN="Reinstall Joomla core files"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLED="Installed Joomla version"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLUPDATE="Install the Update"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLUPDATE="Start update process"
```
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLUPDATE="Update"
```

brianteeman commented on administrator/language/en-GB/com_joomlaupdate.ini:
```diff
@@ -15,26 +16,45 @@ COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_NEXT="Joomla Next"
 COM_JOOMLAUPDATE_CONFIGURATION="Joomla Update: Options"
 COM_JOOMLAUPDATE_CONFIRM="Confirm"
+COM_JOOMLAUPDATE_EMPTYSTATE_APPEND="Update your site by uploading the update package manually."
+COM_JOOMLAUPDATE_EMPTYSTATE_BUTTON_ADD="Check for updates"
```
```suggestion
COM_JOOMLAUPDATE_EMPTYSTATE_BUTTON_ADD="Check for Updates"
```

bembelimen committed:
Update administrator/components/com_joomlaupdate/src/View/Joomlaupdate/HtmlView.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Update administrator/components/com_installer/src/Model/WarningsModel.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/language/en-GB/com_joomlaupdate.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/update.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/reinstall.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Update administrator/components/com_joomlaupdate/tmpl/joomlaupdate/noupdate.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Update administrator/components/com_joomlaupdate/src/View/Upload/HtmlView.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Set correct memory limit

bembelimen committed:
Merge branch '4.0/pre-updatecheck' of github.com:bembelimen/joomla-cms into 4.0/pre-updatecheck

# Conflicts:
#	administrator/components/com_installer/src/Model/WarningsModel.php

bembelimen committed:
Update language information

Quy commented on administrator/components/com_installer/src/Model/WarningsModel.php:
```diff
@@ -95,7 +95,11 @@ public function getItems()
-		$messages = array();
+		$messages = [];
+
+		 // 16MB
+		$minMemory = 16 * 1024 * 1024;
```
```suggestion
		// 16MB
		$minMemory = 16 * 1024 * 1024;
```

bembelimen committed:
Update administrator/components/com_installer/src/Model/WarningsModel.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Improve details button

bembelimen committed:
Merge branch '4.0/pre-updatecheck' of github.com:bembelimen/joomla-cms into 4.0/pre-updatecheck

chmst commented:
I have tested this item :white_check_mark: successfully on 005fdd13da259abff415833c3618466c6ddd4ad3<br /><br />Tested following the testing instructions. 

To complete the test, the  layout layouts/joomla/content/emptystate.php is needed. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34754">issues.joomla.org/tracker/joomla-cms/34754</a>.</sub>

Quy commented:
Using custom URL: `https://update.joomla.org/core/nightlies/next_major_list.xml`

![34754-quick-icon](https://user-images.githubusercontent.com/368084/125357086-ff31ad80-e31b-11eb-8531-2ce48922813d.jpg)

Clicking the icon goes to this screen. Clicking `Check for Updates` button display `Checked for updates.` message.
![34754-check-updates](https://user-images.githubusercontent.com/368084/125357093-01940780-e31c-11eb-9b68-5ebb2fb19170.jpg)

bembelimen commented:
> 
> 
> Using custom URL: `https://update.joomla.org/core/nightlies/next_major_list.xml`
> 
> ![34754-quick-icon](https://user-images.githubusercontent.com/368084/125357086-ff31ad80-e31b-11eb-8531-2ce48922813d.jpg)
> 
> Clicking the icon goes to this screen. Clicking `Check for Updates` button display `Checked for updates.` message.
> ![34754-check-updates](https://user-images.githubusercontent.com/368084/125357093-01940780-e31c-11eb-9b68-5ebb2fb19170.jpg)

Hi thanks for the feedback, that is "intended" (or not different to the old version): https://github.com/joomla/joomla-cms/pull/34754#issuecomment-877906513

But probably we should change the title to "your current version is xxx" instead of "check if an update is available".

Quy commented:
Before PR, I got this:

![34754-pre-update-check](https://user-images.githubusercontent.com/368084/125359488-3190da00-e31f-11eb-8686-cbd6b01fd072.jpg)

Quy commented:
Plus the quickicon is misleading giving the impression that there is an update to perform.

bembelimen commented:
Yeah, but can you rewrite the files in the 2nd tab?

Quy commented:
![34754-reinstall](https://user-images.githubusercontent.com/368084/125360369-9e58a400-e320-11eb-912d-feedcff16739.jpg)

bembelimen commented:
I'm using:
![grafik](https://user-images.githubusercontent.com/1229869/125360828-b9510580-e36c-11eb-90ad-89d2e8451e28.png)

and get:
![grafik](https://user-images.githubusercontent.com/1229869/125360873-ca9a1200-e36c-11eb-9a40-712fd75b3f61.png)

Quy commented:
I installed the prebuilt package: https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/34754/downloads/45766/

The custom URL is same as yours but I don't get the same result.

wilsonge commented:
<img width="1280" alt="Screenshot 2021-07-13 at 11 18 07" src="https://user-images.githubusercontent.com/1986000/125435354-1fef8a8b-6773-4f42-a99b-1f89262e747e.png">

Works for me too :/

wilsonge commented:
Thankyou!

particthistle commented:
I'm getting same as @Quy - I too installed the prebuilt package, so that's perhaps the difference there... is it a timing issue on the nightly build? 

![image](https://user-images.githubusercontent.com/5515866/125439541-01498d5f-6cf1-49f6-b562-3a6c10871372.png)

particthistle commented:
But now it's merged, that might make a difference when applied in the next nightly build and RC4.

bembelimen commented:
Especially thanks @brianteeman for the big review and @Quy for the improvements.
Also thanks for the testers

richard67 commented:
Sorry I just see I was wrong regarding the place in script.php where the deleted files should be added.

Not for the reasons which @brianteeman stated above - what I wrote about where things belong to is still right.

But I was missing the fact that in 3.10 these files are at another place (administrator/components/com_joomlaupdate/views/default/tmpl) and so the J3 files are already present in the "From 3.10 to 4.0" section, and the J4 files removed with this PR here should be moved to section "4.0 from RC 3 to RC 4" in the deleted files list in script.php.

I will make a PR to fix that.

brianteeman commented:
lol

richard67 commented:
> lol

@brianteeman Are you perfect so you never make mistakes? I am not, and I am aware of this and able to correct them, in opposite to other people around. And it changes nothing on what I have said about how that deleted files and folder lists are currently structured.

richard67 commented:
See #34784 .
